src/nnet3/nnet-simple-component.cc

```diff
@@ -1011,6 +1022,196 @@ void PerElementScaleComponent::UnVectorize(
+}
+
+std::string PerElementOffsetComponent::Info() const {
+  std::stringstream stream;
+  BaseFloat offsets_mean = offsets_.Sum() / offsets_.Dim();
+  BaseFloat offsets_stddev = std::sqrt(VecVec(offsets_, offsets_) /
+                              offsets_.Dim()) - (offsets_mean * offsets_mean);
+  stream << Type() << ", input-dim=" << InputDim()
+         << ", output-dim=" << OutputDim()
+		 << ", offsets-mean=" << offsets_mean
+         << ", offsets-stddev=" << offsets_stddev
+         << ", learning-rate=" << LearningRate();
+  return stream.str();
+}
+
+Component* PerElementOffsetComponent::Copy() const {
+  PerElementOffsetComponent *ans = new PerElementOffsetComponent();
+  ans->learning_rate_ = learning_rate_;
+  ans->offsets_ = offsets_;
+  ans->is_gradient_ = is_gradient_;
+  return ans;
+}
+
+BaseFloat PerElementOffsetComponent::DotProduct(
+    const UpdatableComponent &other_in) const {
+  const PerElementOffsetComponent *other =
+      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
+  return VecVec(offsets_, other->offsets_);
+}
+
+void PerElementOffsetComponent::Init(
+    BaseFloat learning_rate, int32 dim,
+    BaseFloat param_mean, BaseFloat param_stddev) {
+  UpdatableComponent::Init(learning_rate);
+  offsets_.Resize(dim);
+  KALDI_ASSERT(dim > 0 && param_stddev >= 0.0);
+  offsets_.SetRandn();
+  offsets_.Scale(param_stddev);
+  offsets_.Add(param_mean);
+}
+
+void PerElementOffsetComponent::Init(BaseFloat learning_rate,
+                                    std::string matrix_filename) {
+  UpdatableComponent::Init(learning_rate);
+  CuMatrix<BaseFloat> mat;
+  ReadKaldiObject(matrix_filename, &mat); // will abort on failure.
+  KALDI_ASSERT(mat.NumCols() == 1);
+  int32 dim = mat.NumRows();
+  offsets_.Resize(dim);
+  offsets_.CopyColFromMat(mat, 0);
+}
+
+void PerElementOffsetComponent::InitFromConfig(ConfigLine *cfl) {
+  bool ok = true;
+  BaseFloat learning_rate = learning_rate_;
+  std::string matrix_filename;
+  int32 dim = -1;
+  cfl->GetValue("learning-rate", &learning_rate); // optional.
+  if (cfl->GetValue("matrix", &matrix_filename)) {
+    Init(learning_rate, matrix_filename);
+    if (cfl->GetValue("dim", &dim))
+      KALDI_ASSERT(dim == InputDim() &&
+                   "input-dim mismatch vs. matrix.");
+  } else {
+    ok = ok && cfl->GetValue("dim", &dim);
+    BaseFloat param_mean = 1.0, param_stddev = 0.0;
+    cfl->GetValue("param-mean", &param_mean);
+    cfl->GetValue("param-stddev", &param_stddev);
+    Init(learning_rate, dim, param_mean, param_stddev);
+  }
+  if (cfl->HasUnusedValues())
+    KALDI_ERR << "Could not process these elements in initializer: "
+              << cfl->UnusedValues();
+  if (!ok)
+    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+}
+
+void PerElementOffsetComponent::Propagate(
+    const ComponentPrecomputedIndexes *indexes,
+    const CuMatrixBase<BaseFloat> &in,
+    CuMatrixBase<BaseFloat> *out) const {
+  
+  out->CopyRowsFromVec(offsets_);
```

Comment:
To test it you need to modify GenerateRandomComponentConfig() in
nnet-test-utils.cc.  Make sure you modify the RandInt statement to include
the integer in the case statement you add.
Dan


On Sun, Nov 1, 2015 at 5:18 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43588535>:
>
> > +    cfl->GetValue("param-stddev", &param_stddev);
> > +    Init(learning_rate, dim, param_mean, param_stddev);
> > +  }
> > +  if (cfl->HasUnusedValues())
> > +    KALDI_ERR << "Could not process these elements in initializer: "
> > +              << cfl->UnusedValues();
> > +  if (!ok)
> > +    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
> > +}
> > +
> > +void PerElementOffsetComponent::Propagate(
> > +    const ComponentPrecomputedIndexes *indexes,
> > +    const CuMatrixBase<BaseFloat> &in,
> > +    CuMatrixBase<BaseFloat> *out) const {
> > +
> > +  out->CopyRowsFromVec(offsets_);
>
> Also you can directly use the AddRowsVec() method
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43588535>.
>